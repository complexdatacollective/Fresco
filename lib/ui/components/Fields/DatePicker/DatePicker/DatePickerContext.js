import React from 'react';

const DatePickerContext = React.createContext({
  onChange: () => {},
  range: null,
  date: {},
  type: null,
});

export default DatePickerContext;
