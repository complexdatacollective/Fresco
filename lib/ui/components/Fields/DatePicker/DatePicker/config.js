/* eslint-disable import/prefer-default-export */

export const DEFAULT_MIN_DATE = { years: 100 }; // DateTime.minus(DEFAULT_MIN_DATE);

export const DEFAULT_TYPE = 'full';

export const DATE_FORMATS = {
  full: 'yyyy-MM-dd',
  month: 'yyyy-MM',
  year: 'yyyy',
};
