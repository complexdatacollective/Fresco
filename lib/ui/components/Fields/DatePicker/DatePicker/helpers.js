/* eslint-disable import/prefer-default-export */

import { DateTime, Info } from 'luxon';
import { isEqual, get } from 'lodash';

export const now = () => DateTime.local();

/**
 * Is date object fully complete?
 */
export const isComplete = (type) => ({ day, month, year }) => {
  switch (type) {
    case 'year':
      return !!year;
    case 'month':
      return !!year && !!month;
    default:
      return !!year && !!month && !!day;
  }
};

/**
 * Is date object empty
 */
export const isEmpty = () => (date) => isEqual(date, { year: null, month: null, day: null });

export const getMonthName = (numericMonth) => get(Info.months(), numericMonth - 1, numericMonth);

export const formatRangeItem = (value, props = {}) => ({
  value,
  label: value,
  ...props,
});
