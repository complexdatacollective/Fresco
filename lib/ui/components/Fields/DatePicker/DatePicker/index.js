export { default as Years } from './Years';
export { default as Months } from './Months';
export { default as Days } from './Days';
export { default as Date } from './Date';
export { default as DatePicker } from './DatePicker';
export { DATE_FORMATS } from './config';
export { now } from './helpers';
