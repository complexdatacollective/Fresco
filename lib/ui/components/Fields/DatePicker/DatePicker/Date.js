import { useContext } from 'react';
import DatePickerContext from './DatePickerContext';
import { isComplete, isEmpty } from './helpers';

const Date = ({ children }) => {
  const {
    onChange, date, range, type,
  } = useContext(DatePickerContext);

  return children({
    onChange,
    date,
    range,
    type,
    isComplete: isComplete(type)(date),
    isEmpty: isEmpty(type)(date),
  });
};

export default Date;
