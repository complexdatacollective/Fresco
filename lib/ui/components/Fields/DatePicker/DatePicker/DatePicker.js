import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { DateTime, Interval } from 'luxon';
import DatePickerContext from './DatePickerContext';
import { DATE_FORMATS, DEFAULT_TYPE, DEFAULT_MIN_DATE } from './config';
import { now, isComplete, isEmpty } from './helpers';

/**
 * Get date object from an ISO string
 */
const getDate = (dateString) => {
  const { year, month, day } = dateString
    ? DateTime.fromISO(dateString).toObject()
    : {
      month: null,
      day: null,
      year: null,
    };
  return { year, month, day };
};

const DatePicker = ({
  children,
  date,
  min,
  max,
  onChange,
  type,
}) => {
  const [pickerState, setPickerState] = useState({
    date: getDate(date),
  });

  // Correctly update component state when passed new date prop
  useEffect(() => {
    setPickerState((state) => ({
      ...state,
      date: getDate(date),
    }));
  }, [date]);

  const typeWithDefault = type || DEFAULT_TYPE;

  const format = DATE_FORMATS[typeWithDefault];

  const minWithDefault = min
    ? DateTime.fromISO(min)
    : now().minus(DEFAULT_MIN_DATE);

  const maxWithDefault = max
    ? DateTime.fromISO(max)
    : now();

  const range = Interval.fromDateTimes(minWithDefault.startOf('day'), maxWithDefault.endOf('day'));

  const handleOnChange = (values) => {
    const newDate = { ...pickerState.date, ...values };

    setPickerState((state) => ({
      ...state,
      date: newDate,
    }));

    if (isEmpty(type)(newDate)) {
      onChange('');
      return;
    }

    if (isComplete(type)(newDate)) {
      const dateString = DateTime.fromObject(newDate).toFormat(format);
      onChange(dateString);
    }
  };

  const context = {
    onChange: handleOnChange,
    range,
    type,
    ...pickerState,
  };

  return (
    <DatePickerContext.Provider value={context}>
      {children}
    </DatePickerContext.Provider>
  );
};

DatePicker.defaultProps = {
  children: null,
  type: DEFAULT_TYPE,
  date: null,
  min: null,
  max: null,
  onChange: () => {},
};

DatePicker.propTypes = {
  children: PropTypes.node,
  date: PropTypes.string,
  type: PropTypes.oneOf([...Object.keys(DATE_FORMATS)]),
  min: PropTypes.string,
  max: PropTypes.string,
  onChange: PropTypes.func,
};

export default DatePicker;
