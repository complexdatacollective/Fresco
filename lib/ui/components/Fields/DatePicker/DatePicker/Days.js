import { useContext } from 'react';
import { range } from 'lodash';
import { DateTime } from 'luxon';
import DatePickerContext from './DatePickerContext';
import { formatRangeItem } from './helpers';

/**
 * Supplies `days` range based on currently selected month.
 */
const Days = ({ children }) => {
  const { date, range: dateRange } = useContext(DatePickerContext);

  const days = range(1, DateTime.fromObject(date).daysInMonth + 1)
    .map((day) => {
      const d = DateTime.fromObject({ ...date, day });
      if (dateRange.contains(d)) {
        return formatRangeItem(day);
      }
      return formatRangeItem(day, { isOutOfRange: true });
    });

  return children({ days });
};

export default Days;
