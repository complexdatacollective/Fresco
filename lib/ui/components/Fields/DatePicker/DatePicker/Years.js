import { useContext } from 'react';
import { range } from 'lodash';
import DatePickerContext from './DatePickerContext';
import { formatRangeItem } from './helpers';

/**
 * Supplies `years` range based on min/max props.
 */
const Years = ({ children }) => {
  const { range: dateRange } = useContext(DatePickerContext);
  const years = range(dateRange.start.year, dateRange.end.year + 1)
    .map((y) => formatRangeItem(y));
  return children({ years });
};

export default Years;
