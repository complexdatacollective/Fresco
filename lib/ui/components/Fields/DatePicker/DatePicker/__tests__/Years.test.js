/* eslint-env jest */

import React from 'react';
import { mount } from 'enzyme';
import DatePicker from '../DatePicker';
import Years from '../Years';

const getSubject = (props) => {
  const mockFunctionalComponent = jest.fn(() => null);

  mount((
    // eslint-disable-next-line react/jsx-props-no-spreading
    <DatePicker {...props}>
      <Years>{mockFunctionalComponent}</Years>
    </DatePicker>
  ));

  return mockFunctionalComponent;
};

describe('<Years>', () => {
  it('provides years from min to max', () => {
    const subject = getSubject({
      min: '1977-07-07',
      max: '1989-09-09',
    });

    expect(subject.mock.calls[0][0])
      .toMatchSnapshot();
  });
});
