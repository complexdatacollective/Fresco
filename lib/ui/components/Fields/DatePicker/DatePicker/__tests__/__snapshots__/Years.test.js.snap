// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Years> provides years from min to max 1`] = `
Object {
  "years": Array [
    Object {
      "label": 1977,
      "value": 1977,
    },
    Object {
      "label": 1978,
      "value": 1978,
    },
    Object {
      "label": 1979,
      "value": 1979,
    },
    Object {
      "label": 1980,
      "value": 1980,
    },
    Object {
      "label": 1981,
      "value": 1981,
    },
    Object {
      "label": 1982,
      "value": 1982,
    },
    Object {
      "label": 1983,
      "value": 1983,
    },
    Object {
      "label": 1984,
      "value": 1984,
    },
    Object {
      "label": 1985,
      "value": 1985,
    },
    Object {
      "label": 1986,
      "value": 1986,
    },
    Object {
      "label": 1987,
      "value": 1987,
    },
    Object {
      "label": 1988,
      "value": 1988,
    },
    Object {
      "label": 1989,
      "value": 1989,
    },
  ],
}
`;
